### PR TITLE
toccata: add MacroSystem Toccata AD1848 sound board support (AHI, playback)

### PR DESCRIPTION
**Pull request by sidick**
Addresses #461.

**Depends on #463** (still open): this branch is stacked on top of add-toccata-audio-support, so the diff below currently includes #463's audio-sink/stem-capture commits as well as the Toccata-specific ones. Once #463 merges, GitHub will recalculate this PR's diff against the updated `main` and it will shrink to just the 7 Toccata commits. Please review/merge #463 first; #464 doesn't need to be held for anything else.

## Summary

M1–M3 of a staged Toccata implementation (M0, the host-side audio sink/stem-capture service this plugs into, is #463): a register-accurate MacroSystem Toccata sound board, so any AHI-aware Amiga application gets 16-bit sound through the stock, open-source `toccata.audio` AHI driver with no Copperline-specific guest software.

Modelled against WinUAE/amiberry's `sndboard.cpp` (byte-identical in both trees) as a behavioural oracle — register offsets, the FIFO's threshold/byte-order quirks, and the interrupt condition are transcribed from what that reference does, not reconstructed from the AD1848 datasheet alone.

- **`src/toccata/ad1848.rs`** — the codec core: 16 indexed registers, 1024-byte play FIFO with an edge-triggered half-empty latch, the board's status/control register (including its RESET-zeroes-the-byte and write-exactly-`0x01`-flushes-the-FIFO idioms), reg 12 pinned to `0x0A` (plain-AD1848 mode only — no CS4231 extensions, no µ-law/A-law), the 14-rate crystal/divider table, DAC output volume, and the documented real-hardware quirk where a native `move.w` write is byte-swapped by the FIFO's little-endian readback.
- **`src/toccata.rs`** — the board wrapper: autoconfig (manufacturer 18260/MacroSystem, product 12, 64 KB Zorro II window, no autoboot ROM), address decode by bit pattern (heavily mirrored, matching the reference), and the mixer-rate cadence.
- **Mixer integration** — a new `ToccataAudioRing` alongside the existing `CdAudioRing`, a generalized `DeviceHost::for_slot_with_audio` carrying both rings to the bus's board-tick loop, and a fifth `push_mixed_frame` tap. The codec's own programmed rate (independent of the mixer's fixed 44.1 kHz) is converted through a polyphase windowed-sinc resampler — promoted out of `src/mt32/` into `src/audio/resample.rs` so both consumers share it — cached per codec rate so switching back to an already-used rate never rebuilds its kernel table.
- **`[toccata] enabled = true`** config, board attaches in `build_machine`, and the board registers as a `toccata` audio source for `--audio-stems`.
- Docs: new `docs/internals/toccata.md`, cross-linked from `docs/zorro.md`, `docs/guide/configuration.md`, and `docs/internals/audio.md`.

## Scope

In scope: playback (8-bit unsigned and 16-bit signed linear PCM, mono/stereo, all 14 AD1848 rates), autoconfig recognition, and INT6 interrupt semantics.

Out of scope for this PR, both explicitly noted in the docs and commit messages:
- **Record** — the record FIFO/interrupt path exists as inert stubs (never activates, since nothing sets `STATUS_FIFO_RECORD`).
- **The board's own "Paula/CD audio mixer" AUX input gain** — Copperline's Paula and CD-DA already reach the master mix directly, so modelling the board's internal analog mixer would add no user-visible capability.
- **Launcher GUI support** — `[toccata]` works via `--config`/a TOML file, but the launcher's machine-configuration screen doesn't yet have a Toccata row, so loading such a config into the launcher and saving from there would currently drop the setting (the same failure class the M0 PR's review caught and fixed for `[audio] stem_granularity`).
- **Full interactive AHI-driver verification** — the assets needed (AHI 4.18 + `toccata.audio`, already present locally) are documented in `tests/README.md`'s new recipe, and an asset-gated test (`toccata_ahi_driver_recognizes_the_board`) is scaffolded and ready, but actually building the HDF and confirming AHI prefs lists the board's modes is left as a follow-up better done with iterative screenshot review than guessed at autonomously.

## Test plan

- [x] `cargo test` — 2653 lib tests pass, including 30 new Toccata-specific unit tests (register semantics, FIFO edge cases, autoconfig identity, mixer-cadence end-to-end at the codec's native 44.1 kHz rate)
- [x] `cargo test --release --test probe_golden` — 28/28 golden renders pass (no video/timing regression)
- [x] `cargo test --test toccata -- --ignored` — the asset-free test passes for real (board autoconfigs, `toccata.wav` stem is written); the asset-gated AHI test skips cleanly without local assets
- [x] `cargo test --test audio_stems_determinism -- --ignored` — still green alongside the new board
- [x] `cargo fmt --check`, `cargo clippy --all-targets --all-features --locked -- -D warnings` — clean
- [x] `cargo check --target wasm32-unknown-unknown` (crates/copperline-web) — clean
- [x] Manual: booted with `[toccata] enabled = true`, confirmed the board autoconfigs at a real Zorro II address in the log, and `--audio-stems-mode master,source,channel` writes `toccata.wav` alongside every other source

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_019j8jbZX3Z2kZPYf6M2UVzg

### CODE REVIEW

File: src/toccata.rs [hunk 1]
```diff
@@ -0,0 +1,331 @@
...
+            // pulls from via the refill closure below.
+            let Self {
+                ad1848, resamplers, ..
+            } = self;
+            let rate = ad1848.sample_rate_hz();
+            let resampler = resamplers
+                .entry(rate)
+                .or_insert_with(|| Resampler::new(rate, MIX_SAMPLE_RATE));
+            let (left, right) = resampler.next(|| ad1848.produce_one_sample());
+            ring.push_frame(left, right);
```
[the remark this answers is not in the document]
Comment (sidick) [reply src/toccata.rs]: Fixed in b7589ab: split into two independent exact-ratio accumulators. `advance_codec` paces `produce_one_sample()` causally at the codec's own rate into a passive `decoded` queue; `advance_mixer` only ever pulls from that queue (or repeats the chip's last sample) through the resampler, never calling back into the chip from inside the resampler's own refill closure. New regression test: `advance_codec_does_not_drain_the_fifo_before_its_own_cadence_elapses`.

File: src/toccata.rs [hunk 2]
```diff
@@ -0,0 +1,331 @@
...
+    /// Mixer-rate accumulator, in units of colour-clocks * MIX_SAMPLE_RATE
+    /// -- the exact same accumulator shape `Paula::advance_audio` uses.
+    /// One output frame is due each time this reaches PAULA_CLOCK_HZ.
+    mixer_acc: u64,
+    /// One resampler per codec rate this session has used (at most 14, the
+    /// AD1848's legal rate count), so switching back to an already-seen
+    /// rate never rebuilds its kernel table. Purely a host-side cache, not
+    /// machine state -- rebuilt lazily after a savestate load.
+    #[serde(skip)]
+    resamplers: HashMap<u32, Resampler>,
```
[the remark this answers is not in the document]
Comment (sidick) [reply src/toccata.rs:103]: Fixed in b7589ab: rather than accept the discontinuity, the resampler cache is now genuinely persisted. `Resampler` gained manual `Serialize`/`Deserialize` impls (in `src/audio/resample.rs`) that carry `l`/`m`/`phase`/`history`/`head`/`primed` and rebuild the derived kernel table from `l`/`m` on load rather than storing it. `Toccata::resamplers` is no longer `#[serde(skip)]`. New test `savestate_round_trip_reproduces_an_uninterrupted_runs_output` proves a state saved and reloaded mid-stream produces exactly the frames an uninterrupted run would have.

File: src/toccata/ad1848.rs [hunk 3]
```diff
@@ -0,0 +1,680 @@
...
+        if v & 0x02 != 0 {
+            self.active |= STATUS_FIFO_RECORD;
+        }
+        self.regs[9] = v;
+    }
+
+    fn codec_start(&mut self) {
+        // Format/rate take effect from here; a driver programs reg 8
+        // before enabling the codec via reg 9, matching hardware.
+    }
```
[the remark this answers is not in the document]
Comment (sidick) [reply src/toccata/ad1848.rs:261]: Fixed in b7589ab: `Ad1848` now latches `active_channels`/`active_sixteen_bit`/`active_rate_hz` at `codec_start()` (reg 9's stopped-to-started transition) via a shared `decode_format()` helper, instead of `channels()`/`sixteen_bit()`/`sample_rate_hz()` reading register 8 live. New test `reprogramming_reg8_while_active_does_not_change_the_active_frame_width` proves format stays fixed mid-stream until the next start.

File: src/main.rs [hunk 4]
```diff
@@ -2214,6 +2324,23 @@ fn main() -> Result<()> {
...
+                anyhow!(
+                    "--audio-stems requires --audio-stems-mode LIST (e.g. \"master,source\"), \
+                     or a [audio] stem_granularity default in the config"
+                )
+            })?;
+        let sources = configured_audio_stem_sources(&cfg);
+        emu.bus_mut()
+            .paula
+            .audio
+            .enable_stems(dir, granularities, &sources)?;
```
[the remark this answers is not in the document]
Comment (sidick) [reply src/main.rs]: Fixed in 899ad37, which merges in PR #463's `state_loaded`-aware `configured_audio_stem_sources` (the two PRs shared this exact concern for CD/MT-32) and extends the same treatment to `toccata`: the call site was already correctly placed after `--load-state` runs (`emu.load_state`), so the only gap was that `toccata`'s registration checked `cfg.toccata` unconditionally instead of `state_loaded || cfg.toccata` like the CD/MT-32 checks. Test `stem_sources_are_conservative_after_a_state_load` now asserts `toccata` is registered under `state_loaded=true` regardless of `cfg`, and absent otherwise.

**Review by LinuxJedi — APPROVED** [review]
